13_sandboxes/sandbox_pool.py

```diff
@@ -0,0 +1,231 @@
+HEALTH_CHECK_TIMEOUT_SECONDS = 10
+SANDBOX_TIMEOUT_SECONDS = 5 * 60
+SANDBOX_USE_DURATION_SECONDS = 2 * 60
+
+
+# ## Health check
+#
+# In this example, we run a very simple health check that just ensures that the
+# server is running and responding to requests.
+def health_check_sandbox(daemon_url):
+    import requests
+
+    start_time = time.time()
+    while time.time() - start_time < HEALTH_CHECK_TIMEOUT_SECONDS:
+        try:
+            response = requests.get(daemon_url, timeout=HEALTH_CHECK_TIMEOUT_SECONDS)
+            response.raise_for_status()
+            return
+        except requests.RequestException:
+            if time.time() - start_time >= HEALTH_CHECK_TIMEOUT_SECONDS:
+                raise
+            time.sleep(0.1)
+
+    raise requests.RequestException("Health check failed")
+
+
+# ## Adding a sandbox to the pool
+#
+# This function is called by the `resize_pool` function to add a sandbox to the pool,
+# and also by the `get_sandbox` function to replace the sandbox that was claimed.
+#
+# It creates a new sandbox, runs the health check, and adds the sandbox to the pool.
+@app.function(image=server_image, retries=3)
+@modal.concurrent(max_inputs=100)
+def add_sandbox_to_queue():
+    # This is done so we don't create sandboxes in the ephemeral app
+    deployed_app = modal.App.lookup("sandbox-pool", create_if_missing=True)
+
+    server_command = ["python", "-m", "http.server", f"{SERVER_PORT}"]
+    sb = modal.Sandbox.create(
+        *server_command,
+        app=deployed_app,
+        image=sandbox_image,
+        encrypted_ports=[SERVER_PORT],
+        timeout=SANDBOX_TIMEOUT_SECONDS,
+    )
+    expiration_time = int(time.time()) + SANDBOX_TIMEOUT_SECONDS
+    url = sb.tunnels()[SERVER_PORT].url
+
+    health_check_sandbox(url)
+
+    pool_queue.put((sb.object_id, url, expiration_time))
+
+
+# ## Getting a sandbox from the pool
+#
+# This is deployed as a web endpoint to claim a sandbox from the pool.
+#
+# It checks the pool for a sandbox that has enough time left, and returns the URL
+# to the server running in the sandbox.
+@app.function(image=server_image)
+@modal.fastapi_endpoint()
+@modal.concurrent(max_inputs=100)
+def get_sandbox() -> str:
+    while True:
+        # backfill + ensures the queue will have at least one sandbox
+        add_sandbox_to_queue.spawn()
+
+        sb_id, url, expiration_time = pool_queue.get(timeout=None)
+
+        if expiration_time < time.time() + SANDBOX_USE_DURATION_SECONDS:
+            print(f"Sandbox {sb_id} does not have enough time left - removing it")
+            sb = modal.Sandbox.from_id(sb_id)
+            sb.terminate()
+            continue
+
+        return url
+
+
+# ## Resizing the pool
+#
+# This function grows or shrinks the pool to the desired size.
+#
+# It can be called programmatically, or manually by running e.g.
+# `modal run 13_sandboxes/sandbox_pool.py::resize_pool --target 3`
+@app.local_entrypoint()
+def resize_pool(target: int = 2):
```

Comment:
It feels to me like this function has the potential to race badly with calls to `get_sandbox`. I would advocate for an approach that doesn't treat the size of the pool as the source of truth for the desired size. I think it would make more sense to store some `desired_size` in a dict and call a function in a cron which reconciles towards that size. I have this bad feeling that this implementation has the potential for drift away from the desired size.

Reply:
Yes agreed, I left it out to keep the example as simple as possible, but there's of course a tradeoff. The idea in a demo setting is that you `resize_pool(3)` when deploying the app, and `resize_pool(0)` when shutting things down, and then the pool refills itself as sandboxes are claimed, so it should never go _below_ the `target` size (though might occasionally go above).

I'd be curious to hear @charlesfrye's thoughts too on what makes sense in this examples repo context, or maybe this should rather live somewhere else.